scipy/sparse/csgraph/_flow.pyx

```diff
@@ -670,3 +699,875 @@ cdef ITYPE_t[:] _dinic(
+        if the input graph is not in CSR format.
+
+    ValueError:
+        If any of the following conditions is met:
+          - the capacity values, vertex demands or edge costs are not integers
+          - an edge cost or vertex demand are infinite in absolute value
+          - an edge has negative capacity
+          - the sum of demands is not zero
+          - there doesn't exist a flow which satisfies all vertex demands.
+
+    Notes
+    -----
+    This solves the minimum cost flow problem [1]_ on a given directed weighted graph:
+    a flow associates to every edge a value, also called a flow, less than the
+    capacity of the edge, and for provided cost of the edge so that for every
+    vertex (apart from the source and the sink vertices), the difference between
+    the total incoming flow and the total outgoing flow is equal to the demand
+    of that vertex. The value of a flow is the sum of the flow on all edges and
+    the minimum cost flow problem consists of finding a flow whose cost (which is
+    defined as the scalar product of the edges' flows and costs) is minimal.
+
+    To solve the problem, we use Network Simplex [2]_.
+    The time complexity of the former :math:`O(|V|^2\,|E|\,log(|V|\,C))`.
+
+    The minimum cost flow problem is usually defined with real valued capacities,
+    edge costs and vertex demands but we require that all capacities are integral.
+    When dealing with rational capacities, or capacities belonging to
+    :math:`x\mathbb{Q}` for some fixed :math:`x \in \mathbb{R}`, it is possible
+    to reduce the problem to the integral case by scaling all capacities
+    accordingly.
+
+    Solving a minimum cost flow problem has many practical uses such as
+    in assignment problem, covers and matching in bipartite graphs [3]_.
+
+    We have ported the Python implementation of NetworkX [4]_ to Cython.
+
+    References
+    ----------
+
+    .. [1] https://en.wikipedia.org/wiki/Minimum-cost_flow_problem#Definition
+    .. [2] Orlin, J.B. A polynomial time primal network simplex algorithm
+           for minimum cost flows. Mathematical Programming 78,
+           109-129 (1997).
+    .. [3] https://en.wikipedia.org/wiki/Network_simplex_algorithm#Applications
+    .. [4] A Hagberg, D Schult, P Swart, Exploring Network Structure, Dynamics,
+           and Function using NetworkX in Proceedings of the 7th Python in
+           Science conference (SciPy 2008), G Varoquaux, T Vaught, J Millman
+           (Eds.), pp. 11-15
+
+    Examples
+    --------
+    >>> from scipy.sparse import csr_matrix
+    >>> from scipy.sparse.csgraph import minimum_cost_flow
+    >>> A = csr_matrix([[0, 4, 10, 0],
+    ...                 [0, 0, 0, 9],
+    ...                 [0, 0, 0, 5],
+    ...                 [0, 0, 0, 0]])
+    >>> demand = [-5, 0, 0, 5]
+    >>> cost = [3, 6, 1, 2]
+    >>> result = minimum_cost_flow(A, demand, cost)
+    >>> result.flow_cost
+    24
+    >>> result.residual.toarray()
```

Comment:
After seeing #14731 I wanted to check the naming here as well.

And I'm actually a little confused since it seems like `MinCostFlowResult.residual` is neither the residual or the flow. Indeed, the flow in this case is

```
array([[ 0,  4,  1,  0],
       [-4,  0,  0,  4],
       [-1,  0,  0,  1],
       [ 0, -4, -1,  0]])
```

which is almost what is returned here (except the anti-symmetry isn't included), and the residual of the flow is

```
matrix([[0, 0, 9, 0],
        [4, 0, 0, 5],
        [1, 0, 0, 4],
        [0, 4, 1, 0]])
```

Reply:
In terms of what to actually do, I'd probably rename what's currently called `MinimumCostFlowResult.residual` to `MinimumCostFlowResult.flow`, and not return anything called residual at all as it is trivial to calculate from the flow. And then look into whether the result should be skew-symmetric.

Reply:
Done.

Reply:
It's still not anti-symmetric, right? I.e. I'd expect `result.flow[u, v] == -result.flow[v, u]` for any `u` and `v`.

Reply:
What to return if its not skew-symmetric?

Reply:
My comment just is that it should be skew-symmetric, by definition. From a quick look, where you do `edge_flow[i] += f`, I might expect a similar `-= f` for the reverse edge. But maybe there's a good reason it's done done like that?

Reply:
I don't remember exactly why I didn't do `edge_flow[rev_edge] -= f`. I will try doing this.

Reply:
I have updated the code to return skew-symmetric matrix. Tests have also been updated. Please let me know if anything else is to be done.